pkg/controllers/allocation/batch.go

```diff
@@ -0,0 +1,178 @@
+	return &Batcher{
+		MaxBatchPeriod: maxBatchPeriod,
+		IdlePeriod:     idlePeriod,
+		windows:        map[string]*window{},
+	}
+}
+
+// Start should be called before Add or Wait
+// It is not safe to call Start concurrently
+// but Start can be called synchronously multiple times w/ no effect
+func (b *Batcher) Start(ctx context.Context) {
+	if !b.isMonitorRunning {
+		b.ops = make(chan *batchOp, 1000)
+		go b.monitor(ctx)
+		b.isMonitorRunning = true
+	}
+}
+
+// Add starts a batching window or adds to an existing in-progress window
+// Add is safe to be called concurrently
+func (b *Batcher) Add(obj metav1.Object) {
+	select {
+	case b.ops <- &batchOp{kind: opAdd, key: b.keyFrom(obj)}:
+	// Do not block if the channel is full
+	default:
+	}
+}
+
+// Wait blocks until a batching window ends
+// If the batch is empty, it will block until something is added or the window times out
+func (b *Batcher) Wait(obj metav1.Object) {
+	waitBatchOp := &batchOp{kind: opWait, key: b.keyFrom(obj), waitEnd: make(chan bool, 1)}
+	timeout := time.NewTimer(b.MaxBatchPeriod)
+	select {
+	case b.ops <- waitBatchOp:
+		<-waitBatchOp.waitEnd
+	// if the ops channel is full (should be very rare), allow wait to block until the MaxBatchPeriod
+	case <-timeout.C:
+	}
+}
+
+// monitor is a synchronous loop that controls the window start, update, and end
+// monitor should be executed in one go routine and will handle all object batch windows
+func (b *Batcher) monitor(ctx context.Context) {
+	defer func() { b.isMonitorRunning = false }()
+	ticker := time.NewTicker(b.IdlePeriod / 2)
+	for {
+		select {
+		// Wake and check for any timed out batch windows
+		case <-ticker.C:
+			for key, batch := range b.windows {
+				b.checkForWindowEndAndNotify(key, batch)
+			}
+		// Process window operations
+		case op := <-b.ops:
+			switch op.kind {
+			// Start a new window or update progress on a window
+			case opAdd:
+				b.startOrUpdateWindow(op.key)
+			// Register a waiter and start a window if no window has been started
+			case opWait:
+				window, ok := b.windows[op.key]
+				if !ok {
+					window = b.startOrUpdateWindow(op.key)
+				}
+				window.closed = append(window.closed, op.waitEnd)
+			}
+		// Stop monitor routine on shutdown
+		case <-ctx.Done():
+			for key, window := range b.windows {
+				b.endWindow(key, window)
+			}
+			return
+		}
+	}
+}
+
+// checkForWindowEndAndNotify checks if a window has timed out due to inactivity (IdlePeriod) or has reached the MaxBatchPeriod.
+// If the batch window has ended, then the batch closed channel will be notified and the window will be removed
+func (b *Batcher) checkForWindowEndAndNotify(key string, window *window) {
+	if time.Since(window.lastUpdated) < b.IdlePeriod && time.Since(window.started) < b.MaxBatchPeriod {
+		return
+	}
+	b.endWindow(key, window)
+}
+
+// endWindow signals the end of a window to all wait consumers and deletes the window
+func (b *Batcher) endWindow(key string, window *window) {
```

Comment:
Thought's on calling this close?

Reply:
It used to be close, but it felt awkward to have a window and close and channels closing (since this is private to the Batcher object).

Reply:
Let's revisit some of this naming at some point once I'm more familiar with this code.